pkg/utils/service/service.go

```diff
@@ -0,0 +1,321 @@
+/*
+Package service unifies how our HTTP services run to deduplicate code between different
+microservices. The common Service struct should be embedded into custom server
+implementations and it should be registered with a Server that implements the routes
+and handler functionality. The Service struct handles log initialization, the
+construction and management of the http Server, the construction and management of a
+gin Engine for multiplexing, and the liveness and readiness probes for Kubernetes.
+*/
+package service
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"os"
+	"os/signal"
+	"sync"
+	"time"
+
+	"github.com/gin-gonic/gin"
+	"github.com/hashicorp/go-multierror"
+	"github.com/rotationalio/ensign/pkg/utils/probez"
+	"github.com/rs/zerolog/log"
+)
+
+// Service specifies the methods that a service implementation for a specific
+// microservice needs to implement in order to be registered with the http server. The
+// simplest services only need to specify the Service interface in order to create a
+// quickly working service!
+type Service interface {
+	// Routes is called once the server is up and running and is in a live state but
+	// before it is in a ready state. The intended use case of this method is for the
+	// Service to specify its own middleware, routes, and handlers on the router, which
+	// is the majority of what differentiates a service from the main server.
+	// If an error is returned, the error will shutdown the server and stop startup.
+	Routes(router *gin.Engine) error
+
+	// Shutdown is called after the http server has gracefully concluded serving and the
+	// service is in a not ready, not healthy state. This method should be used to
+	// clean up connections to databases, close resources properly, etc. Note that the
+	// context may have a deadline on it before the server is terminated.
+	// If an error is returned, the error will be reported to the caller of Serve().
+	Shutdown(ctx context.Context) error
```

Comment:
This is creating a bit of confusion with the embedding so I had to rename the `Server` method `GracefulShutdown` - but we need to make sure that users call `GracefulShutdown` and not `Shutdown`. We could rename this `TearDown` or `CleanUp` or `Close` to prevent confusion and rename `GracefulShutdown` back to `Shutdown` - thoughts?